src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseCodeReviewFeedbackService.java

```diff
@@ -0,0 +1,239 @@
+        }
+        finally {
+            unlockRepository(participation, programmingExercise);
+        }
+    }
+
+    /**
+     * Sets an individual due date for a participation, locks the repository,
+     * and invalidates previous results to prepare for new feedback.
+     *
+     * @param participation       the programming exercise student participation.
+     * @param programmingExercise the associated programming exercise.
+     */
+    private ProgrammingExerciseStudentParticipation setIndividualDueDateAndLockRepository(ProgrammingExerciseStudentParticipation participation,
+            ProgrammingExercise programmingExercise, boolean invalidatePreviousResults) {
+        // The participations due date is a flag showing that a feedback request is sent
+        participation.setIndividualDueDate(now());
+
+        participation = programmingExerciseStudentParticipationRepository.save(participation);
+        // Circumvent lazy loading after save
+        participation.setParticipant(participation.getParticipant());
+        programmingExerciseParticipationService.lockStudentRepositoryAndParticipation(programmingExercise, participation);
+
+        if (invalidatePreviousResults) {
+            var participationResults = participation.getResults();
+            participationResults.forEach(participationResult -> participationResult.setRated(false));
+            this.resultRepository.saveAll(participationResults);
+        }
+
+        return participation;
+    }
+
+    /**
+     * Removes the individual due date for a participation. If the due date for an exercise is empty or is in the future, unlocks the repository,
+     *
+     * @param participation       the programming exercise student participation.
+     * @param programmingExercise the associated programming exercise.
+     */
+    private void unlockRepository(ProgrammingExerciseStudentParticipation participation, ProgrammingExercise programmingExercise) {
+        if (programmingExercise.getDueDate() == null || now().isBefore(programmingExercise.getDueDate())) {
+            programmingExerciseParticipationService.unlockStudentRepositoryAndParticipation(participation);
+            participation.setIndividualDueDate(null);
+            this.programmingExerciseStudentParticipationRepository.save(participation);
+        }
+    }
+
+    private void checkRateLimitOrThrow(ProgrammingExerciseStudentParticipation participation) {
+
+        List<Result> athenaResults = participation.getResults().stream().filter(result -> result.getAssessmentType() == AssessmentType.AUTOMATIC_ATHENA).toList();
+
+        long countOfAthenaResultsInProcessOrSuccessful = athenaResults.stream().filter(result -> result.isSuccessful() == null || result.isSuccessful() == Boolean.TRUE).count();
+
+        long countOfSuccessfulRequests = athenaResults.stream().filter(result -> result.isSuccessful() == Boolean.TRUE).count();
+
+        if (countOfAthenaResultsInProcessOrSuccessful >= 3) {
```

Comment:
Would it make sense to make the rate limit configurable rather than hardcoding the threshold?

Reply:
Yes, but in a follow-up PR

Reply:
Fair!